### PR TITLE
Make ioctl be aware of current account in context 1295

### PR DESCRIPTION
**Pull request by lzxm160**
work on #1295

Example:
1、ioctl action transfer aaa 1.5455
Error: use "ioctl config set currentcontext" to config current account first

2、ioctl config set currentcontext aa
currentcontext is set to aa

3、ioctl account balance
io1vdtfpzkwpyngzvx7u2mauepnzja7kd5rryp0sg: 88.884722278 IOTX

4、ioctl account sign "12"

Behaior changed ctl:
**account balance
account delete
account export
account exportpublic
account nonce
account update
account sign
action claim
action deploy
action deposit
action invoke
action read
action transfer
xrc20 approve
xrc20 allowance
xrc20 balanceof
xrc20 transfer
xrc20 transferFrom
alias remove
config set currentcontext
config get currentcontext**

### CODE REVIEW

File: cli/ioctl/cmd/config/config.go [hunk 1]
```diff
@@ -41,10 +41,11 @@ var ConfigCmd = &cobra.Command{
...
 type Config struct {
-	Wallet        string            `yaml:"wallet"`
-	Endpoint      string            `yaml:"endpoint"`
-	SecureConnect bool              `yaml:"secureConnect"`
-	Aliases       map[string]string `yaml:"aliases"`
+	Wallet         string            `yaml:"wallet"`
+	Endpoint       string            `yaml:"endpoint"`
+	SecureConnect  bool              `yaml:"secureConnect"`
+	Aliases        map[string]string `yaml:"aliases"`
+	CurrentContext string            `yaml:"currentcontext"`
```
Comment (zjshen14) [inline cli/ioctl/cmd/config/config.go]: Do you want to define 
```
type Context struct {
    CurrentAddress
}
```
For forward compatibility

File: cli/ioctl/cmd/account/accountbalance.go [hunk 2]
```diff
@@ -33,7 +34,11 @@ var accountBalanceCmd = &cobra.Command{
 
 // balance gets balance of an IoTeX blockchain address
 func balance(args []string) (string, error) {
-	address, err := alias.Address(args[0])
+	addr, err := config.GetAddress(args)
```
Comment (zjshen14) [inline cli/ioctl/cmd/account/accountbalance.go]:
```suggestion
	var addr string
	var err error
	if len(args) == 0 {
	    addr, err = config.GetAddress()
	} else {
	    addr, err = alias.Address(addr)
	}
```

Also,

1. make `config.GetAddress` as simple as fetching the address from config
2. See if the code above could be a common util method

File: cli/ioctl/cmd/config/configsetget.go [hunk 3]
```diff
@@ -110,6 +135,8 @@ func set(args []string) (string, error) {
 		ReadConfig.SecureConnect = !Insecure
 	case "wallet":
 		ReadConfig.Wallet = args[1]
+	case "currentcontext":
+		ReadConfig.CurrentContext = args[1]
```
Comment (zjshen14) [inline cli/ioctl/cmd/config/configsetget.go]: When setting the address, let's try to do some validation. Check ioctl/validator.go

File: cli/ioctl/cmd/alias/aliasremove.go [hunk 4]
```diff
@@ -19,12 +19,16 @@ import (
 
 // aliasRemoveCmd represents the alias remove command
 var aliasRemoveCmd = &cobra.Command{
-	Use:   "remove ALIAS",
+	Use:   "remove [ALIAS]",
```
Comment (zjshen14) [inline cli/ioctl/cmd/alias/aliasremove.go]: IMHO, alias operation doesn't need to be implicit

File: cli/ioctl/cmd/config/configsetget.go [hunk 5]
```diff
@@ -89,9 +92,33 @@ func Get(arg string) (string, error) {
...
+	return currentcontext.AddressOrAlias, nil
+}
+
+// GetAddressOrAlias gets address from args or context
+func GetAddressOrAlias(args []string) (address string, err error) {
+	if len(args) == 1 && !strings.EqualFold(args[0], "") {
+		address = args[0]
+	} else {
+		address, err = GetContextAddressOrAlias()
+		if err != nil {
```
Comment (zjshen14) [inline cli/ioctl/cmd/config/configsetget.go]: No need L115-L117

File: cli/ioctl/cmd/config/config.go [hunk 6]
```diff
@@ -39,12 +39,18 @@ var ConfigCmd = &cobra.Command{
...
 type Config struct {
-	Wallet        string            `yaml:"wallet"`
-	Endpoint      string            `yaml:"endpoint"`
-	SecureConnect bool              `yaml:"secureConnect"`
-	Aliases       map[string]string `yaml:"aliases"`
+	Wallet         string            `yaml:"wallet"`
+	Endpoint       string            `yaml:"endpoint"`
+	SecureConnect  bool              `yaml:"secureConnect"`
+	Aliases        map[string]string `yaml:"aliases"`
+	CurrentContext Context           `yaml:"currentcontext"`
```
Comment (zjshen14) [inline cli/ioctl/cmd/config/config.go]: `yaml:"currentcontext"` -> `yaml:"currentContext"`

File: cli/ioctl/cmd/config/configsetget.go [hunk 7]
```diff
@@ -110,6 +137,13 @@ func set(args []string) (string, error) {
 		ReadConfig.SecureConnect = !Insecure
 	case "wallet":
 		ReadConfig.Wallet = args[1]
+	case "currentcontext":
+		err1 := validator.ValidateAlias(args[1])
```
Comment (zjshen14) [inline cli/ioctl/cmd/config/configsetget.go]:
```suggestion
		if validator.ValidateAlias(args[1]) != nil &&  validator.ValidateAddress(args[1]) != nil {
```

Short circuit

File: cli/ioctl/util/util.go [hunk 8]
```diff
@@ -149,3 +149,31 @@ func ReadSecretFromStdin() (string, error) {
 	}
 	return string(bytePass), nil
 }
+
+// GetAddress get address from address or alias
+func GetAddress(args []string) (addr string, err error) {
+	addr, err = config.GetAddressOrAlias(args)
```
Comment (zjshen14) [inline cli/ioctl/util/util.go]: L155-L163
```suggestion
	if addr, err = config.GetAddressOrAlias(args); err != nil {
	    return
	}
	addr, err = Address(addr)
	return 
```

File: cli/ioctl/cmd/action/actionread.go [hunk 9]
```diff
@@ -60,11 +60,7 @@ func init() {
 
 // read reads smart contract on IoTeX blockchain
 func read(contract address.Address, bytecode []byte) (string, error) {
-	signer := signerFlag.Value().(string)
-	if len(signer) == 0 {
-		signer = defaultSigner
```
Comment (zjshen14) [inline cli/ioctl/cmd/action/actionread.go]: Also please delete defaultSigner

File: cli/ioctl/cmd/alias/aliasremove.go (hunk 4 above)
Comment (lzxm160) [reply cli/ioctl/cmd/alias/aliasremove.go]: ok

File: cli/ioctl/cmd/config/configsetget.go (hunk 7 above)
Comment (lzxm160) [reply cli/ioctl/cmd/config/configsetget.go]: need check both,input could be alias or address,short circuit only check one
if input is address,ValidateAlias will return error,that's not right.
I need ValidateAddress to check if this address is right,only both ValidateAlias and ValidateAddress return error will return error

File: cli/ioctl/util/util.go (hunk 8 above)
Comment (lzxm160) [reply cli/ioctl/util/util.go]: got it

File: cli/ioctl/cmd/action/actionread.go (hunk 9 above)
Comment (lzxm160) [reply cli/ioctl/cmd/action/actionread.go]: this is the old code

**Pull request merged**